### ISSUE TITLE
static data defined in smp/ipi.h

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#635]
include/smp/ipi.h not only declares data, but also defines all IPI related global variables, while the file is included in multiple places. It works because all kernel code is combined into kernel_all_pp_prune.c, but it freaks me out. Is this intentional?

**Comment by axel-h** [comment]
I agree, this should be avoided and it violates a lot of common C best practice rules also. Since there is a `src/smp/ipi.c`, putting them there should easy and we can keep an extern declaration in the header. Can you make a patch an then we can see if the proof run accepts this.
Putting them in `src/model/statedata.c` might also be an option, but it seem this file is not really intended for all global variables per se. Seems it just collected a few things over time that did not have a better place because there was no *.c file.

**Comment by lsf37** [comment]
`statedata.c` is intended for the global data of the kernel, it is definitely not a random accumulation place, and should be the first place to look at for the main global state.

The exception would implementation details of some mechanims. For IPI this is a bit borderline, `ipi.c` might make sense. @kent-mcleod do I have this right?

**Comment by kent-mcleod** [comment]
`ipi.c` is probably best in this case.  statedata.c is usually for state that is also present in the executable spec.  Currently these variables are for implementing the IPI mechanisms in the C kernel and so placing them in `ipi.c` makes sense. Similar to how `gic_v2.c` contains globals related to the GIC C implementation.
